pkg/api/handlers/compat/images_build.go

```diff
@@ -169,14 +170,52 @@ func BuildImage(w http.ResponseWriter, r *http.Request) {
 
 	// convert addcaps formats
 	containerFiles := []string{}
-	if _, found := r.URL.Query()["dockerfile"]; found {
-		var m = []string{}
-		if err := json.Unmarshal([]byte(query.Dockerfile), &m); err != nil {
-			// it's not json, assume just a string
-			m = []string{filepath.Join(contextDirectory, query.Dockerfile)}
+	// If query paramemter `dockerfile` is set and is
+	// on priority never look for `Dockerfile` or
+	// `Containerfile` in the context directory.
+	dockerFileSet := false
+	if utils.IsLibpodRequest(r) && query.Remote != "" {
+		// The context directory could be a URL.  Try to handle that.
+		anchorDir, err := ioutil.TempDir(parse.GetTempDir(), "libpod_builder")
+		if err != nil {
+			utils.InternalServerError(w, err)
+		}
+		tempDir, subDir, err := buildahDefine.TempDirForURL(anchorDir, "buildah", query.Remote)
+		if err != nil {
+			utils.InternalServerError(w, err)
+		}
+		if tempDir != "" {
+			// We had to download it to a temporary directory.
+			// Delete it later.
+			defer func() {
+				if err = os.RemoveAll(tempDir); err != nil {
+					// We are deleting this on server so log on server end
+					// client does not have to worry about server cleanup.
+					logrus.Errorf("Cannot delete downloaded temp dir %q: %s", tempDir, err)
+				}
+			}()
+			contextDirectory = filepath.Join(tempDir, subDir)
+		} else {
+			// Nope, it was local.  Use it as is.
+			absDir, err := filepath.Abs(query.Remote)
+			if err != nil {
+				utils.BadRequest(w, "remote", query.Remote, err)
+			}
+			contextDirectory = absDir
 		}
-		containerFiles = m
 	} else {
+		if _, found := r.URL.Query()["dockerfile"]; found {
+			var m = []string{}
+			if err := json.Unmarshal([]byte(query.Dockerfile), &m); err != nil {
+				// it's not json, assume just a string
+				m = []string{filepath.Join(contextDirectory, query.Dockerfile)}
+			}
+			containerFiles = m
+			dockerFileSet = true
+		}
+	}
+
+	if !dockerFileSet {
```

Comment:
Can you move this block up to where you are setting dockerFileSet?

Reply:
Yes that should work.

Reply:
the block is needed for all the cases when except when `dockerfile` is set and since it modifies the `contextDir` hence it has to be after `query.Remote` processing.